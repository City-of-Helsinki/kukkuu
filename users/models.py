import logging
from datetime import datetime
from typing import TYPE_CHECKING, Optional, Union

from django.conf import settings
from django.contrib.auth import get_user_model
from django.contrib.auth.models import UserManager as OriginalUserManager
from django.core.exceptions import ObjectDoesNotExist
from django.db import models, transaction
from django.db.models import Q
from django.utils import timezone
from django.utils.functional import cached_property
from django.utils.translation import gettext_lazy as _
from guardian.shortcuts import get_objects_for_user
from helsinki_gdpr.models import SerializableMixin
from helusers.models import AbstractUser

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from events.consts import notification_types_that_need_communication_acceptance
from gdpr.consts import CLEARED_VALUE
from gdpr.models import GDPRModel
from languages.models import Language
from projects.models import ProjectPermission

if TYPE_CHECKING:
    from children.models import Child
    from projects.models import Project
    from subscriptions.models import FreeSpotNotificationSubscription
    from verification_tokens.models import VerificationToken

logger = logging.getLogger(__name__)


class UserQuerySet(models.QuerySet):
    def possible_admins(self):
        # This filtering isn't perfect because
        #   1) it is possible there are a few normal users without a Guardian object
        #   2) this prevents using the same user account for Kukkuu UI and Kukkuu admin
        #      which was convenient in dev/testing, and might be a valid case in
        #      production as well in the future
        # but for now this should be easily good enough.
        return self.filter(guardian=None)


# This is needed when using a custom User queryset
class UserManager(
    OriginalUserManager.from_queryset(UserQuerySet),
    SerializableMixin.SerializableManager,
):
    pass


class User(AbstractUser, GDPRModel, SerializableMixin):
    serialize_fields = (
        {"name": "uuid", "accessor": lambda uuid: str(uuid)},
        {"name": "username"},
        {"name": "first_name"},
        {"name": "last_name"},
        {"name": "email"},
        {
            "name": "administered_projects",
            "accessor": lambda projects: [p.serialize() for p in projects],
        },
        {"name": "last_login", "accessor": lambda t: t.isoformat() if t else None},
        {"name": "date_joined", "accessor": lambda t: t.isoformat() if t else None},
        {"name": "guardian"},
    )

    # NOTE: `is_obsolete` is deprecated after the migration process
    # that deactivates the obsoleted users (`0014_deactivate_obsolete_users.py`).
    # When Tunnistamo was changed to a Keycloak service,
    # the field has been used to mark the users that are not anymore
    # accessible, but are still in the database and need to receive
    # notifications (like emails). The users are not deleted, because
    # they might have children or other objects linked to them and
    # it was important that they were not deactivated or deleted,
    # but just set as unaccessible by the auth service migration process.
    #
    # The `is_active` field is used to mark the users that are
    # not anymore able to log in to the system and the `is_obsolete`
    # users should be deactivated at some point, which then makes this
    # field needless as well.
    is_obsolete = models.BooleanField(
        _("obsoleted"),
        null=False,
        default=False,
        help_text=_(
            "Designates whether the user account is obsoleted "
            "and cannot be anymore accessed, "
            "e.g. after an auth service change process "
            "(when Tunnistamo is changed to a Keycloak service)."
        ),
    )

    objects = UserManager()

    gdpr_sensitive_data_fields = ["first_name", "last_name", "email"]

    class Meta:
        verbose_name = _("user")
        verbose_name_plural = _("users")

    def __str__(self):
        return super().__str__() or self.username

    def get_users_projects_with_permission(
        self, permission: ProjectPermission
    ) -> list["Project"]:
        """
        Get all projects where the user has the given permission.
        """
        from projects.models import Project

        return list(get_objects_for_user(self, permission.value, Project))

    @cached_property
    def administered_projects(self) -> list["Project"]:
        return self.get_users_projects_with_permission(ProjectPermission.ADMIN)

    @property
    def projects_with_family_data_access(self) -> list["Project"]:
        return self.get_users_projects_with_permission(ProjectPermission.VIEW_FAMILIES)

    def has_global_permission(self, project_permission: ProjectPermission) -> bool:
        return self.has_perm(project_permission.permission_name)

    def has_project_permission(
        self, project_permission: ProjectPermission, project: "Project"
    ) -> bool:
        return self.has_perm(
            project_permission.value, project
        ) or self.has_global_permission(project_permission)

    def can_administer_project(self, project: "Project") -> bool:
        return project in self.administered_projects

    def can_publish_in_project(self, project: "Project") -> bool:
        return self.has_project_permission(ProjectPermission.PUBLISH, project)

    def can_manage_event_groups_in_project(self, project: "Project") -> bool:
        return self.has_project_permission(
            ProjectPermission.MANAGE_EVENT_GROUPS, project
        )

    def can_send_messages_to_all_in_project(self, project: "Project") -> bool:
        return self.has_project_permission(
            ProjectPermission.SEND_MESSAGE_TO_ALL_IN_PROJECT, project
        )

    def can_view_families_in_project(self, project: "Project") -> bool:
        return self.has_project_permission(ProjectPermission.VIEW_FAMILIES, project)

    def get_active_verification_tokens(
        self, verification_type: Optional[str] = None
    ) -> list["VerificationToken"]:
        """Filter active verification tokens"""
        from verification_tokens.models import VerificationToken

        return VerificationToken.objects.filter_active_tokens(
            self, verification_type=verification_type, user=self
        )

    def deactivate_and_create_email_verification_token(
        self, email: str
    ) -> "VerificationToken":
        from verification_tokens.models import VerificationToken

        return VerificationToken.objects.deactivate_and_create_token(
            self,
            self,
            VerificationToken.VERIFICATION_TYPE_EMAIL_VERIFICATION,
            email,
        )

    def get_subscriptions(
        self, child: Optional["Child"] = None
    ) -> Union[models.QuerySet, list["FreeSpotNotificationSubscription"]]:
        """
        Get all the user's subscriptions.
        If a child argument is given, only the subscriptions that are linked
        to that given child are returned.
        """
        from subscriptions.models import FreeSpotNotificationSubscription

        return FreeSpotNotificationSubscription.objects.user_subscriptions(self, child)

    def unsubscribe_all_notification_subscriptions(
        self, child: Optional["Child"] = None
    ):
        """
        Unsubscribe user's all notification subsriptions.
        If a child argument is given, the unsubscribing is done only
        to subscriptions that are linked to the given child that
        the user is a guardian to.

        NOTE: This function deletes the FreeSpotNotifications, which are linked to
        a Child and Occurrence instances. **It should be noted that the current
        model architecture allows that a child can have multiple guardians,
        so unsubscribe can delete some notifications from other users as well.
        However, the UI apps has never allowed more than 1 guardian for a child.**
        """
        subscriptions = self.get_subscriptions(child=child)
        count = subscriptions.count()
        subscriptions.delete()
        logger.info(
            f"User {self.uuid} unsubscribed all ({count}) "
            "subscriptions they had active%s"
            % (f" for child {child.id}." if child else ".")
        )
        return count

    def create_subscriptions_management_auth_token(self):
        """
        Create a new token that has an expiration date
        and can be used to as an authorization token to manage
        the user's subscriptions to notifications.
        """
        from verification_tokens.models import VerificationToken

        # NOTE: Should the email be left empty?
        # The only value the email gives here is that it is denormalized to db
        try:
            email = self.guardian.email
            if not email:
                email = self.email
        except ObjectDoesNotExist:
            email = self.email

        # NOTE: Should this get_or_create never expiring tokens instead?
        # It would be worse in security perspective, because 1 token would
        # serve forever, but it would be better in usability, because
        # then the token would be the same in every email every time, forever.
        return VerificationToken.objects.create_token(
            self,
            self,
            VerificationToken.VERIFICATION_TYPE_SUBSCRIPTIONS_AUTH,
            email=email,
            expiry_minutes=settings.SUBSCRIPTIONS_AUTH_TOKEN_VALID_MINUTES,
            token_length=settings.SUBSCRIPTIONS_AUTH_TOKEN_LENGTH,
        )

    def clear_gdpr_sensitive_data_fields(self):
        super().clear_gdpr_sensitive_data_fields()
        self.is_active = False
        self.username = f"{CLEARED_VALUE}-{self.uuid}"
        self.set_unusable_password()
        self.save()


class GuardianQuerySet(models.QuerySet):
    def user_can_view(self, user: User):
        return self.filter(
            Q(user=user) | Q(children__project__in=user.administered_projects)
        ).distinct()

    @transaction.atomic()
    def delete(self):
        for child in self:
            child.delete()

    def has_accepted_communication_for_notification(self, notification_type: str):
        """Filters guardians who have accepted communication for the given notification
        type.

        Initially, the queryset is filtered to include only **active**
        (`user__is_active=True`) and **non-obsolete** (`is_obsolete=False`) guardians.
        If the `notification_type` requires explicit acceptance (i.e., it is found
        within `notification_types_that_need_communication_acceptance`), the queryset
        will be further filtered to include only guardians for whom
        `has_accepted_communication` is True.

        Args:
            notification_type (str): The type of notification. This should be one of the
                `NotificationType` constant string literals.

        Returns:
            GuardianQuerySet: A filtered queryset of guardians if the
                `notification_type` requires acceptance; otherwise, returns
                the initially filtered queryset (active and non-obsolete guardians).
        """
        queryset = self.filter(user__is_active=True, user__is_obsolete=False)
        if (
            notification_type
            not in notification_types_that_need_communication_acceptance
        ):
            return queryset
        return queryset.filter(has_accepted_communication=True)

    def for_auth_service_is_changing_notification(
        self,
        user_joined_before: Optional[datetime] = None,
        obsoleted_users_only=True,
        guardian_emails: Optional[list[str]] = None,
    ):
        if user_joined_before and user_joined_before > timezone.now():
            raise ValueError("The user_joined_before cannot be set in future.")

        qs_filters = {"user__date_joined__lte": (user_joined_before or timezone.now())}
        if obsoleted_users_only:
            qs_filters.update({"user__is_obsolete": True})
        if guardian_emails:
            qs_filters.update({"email__in": guardian_emails})
        return self.filter(**qs_filters)


class Guardian(GDPRModel, UUIDPrimaryKeyModel, TimestampedModel, SerializableMixin):
    user = models.OneToOneField(
        get_user_model(), verbose_name=_("user"), on_delete=models.CASCADE
    )
    first_name = models.CharField(verbose_name=_("first name"), max_length=64)
    last_name = models.CharField(verbose_name=_("last name"), max_length=64)
    phone_number = models.CharField(
        verbose_name=_("phone number"), max_length=64, blank=True
    )
    language = models.CharField(
        verbose_name=_("language"), max_length=10, default=settings.LANGUAGE_CODE
    )
    email = models.EmailField(
        _("email address"),
        blank=True,
        help_text=_("If left blank, will be populated with the user's email."),
    )
    languages_spoken_at_home = models.ManyToManyField(
        Language,
        verbose_name=_("languages spoken at home"),
        related_name="guardians",
        blank=True,
    )
    # NOTE: All guardians are initialized to have has_accepted_communication=True
    # in users/migrations/0013_replace_has_accepted_marketing_with_communication.py,
    # before they have implicitly accepted it when they have created an account.
    has_accepted_communication = models.BooleanField(
        _("accepts communication"), null=False, default=False
    )

    serialize_fields = (
        {"name": "id", "accessor": lambda uuid: str(uuid)},
        {"name": "user", "accessor": lambda u: str(u)},
        {"name": "first_name"},
        {"name": "last_name"},
        {"name": "email"},
        {"name": "phone_number"},
        {"name": "has_accepted_communication"},
        {"name": "children"},
    )

    objects = SerializableMixin.SerializableManager.from_queryset(GuardianQuerySet)()

    gdpr_sensitive_data_fields = ["first_name", "last_name", "phone_number", "email"]

    class Meta:
        verbose_name = _("guardian")
        verbose_name_plural = _("guardians")
        ordering = ["-created_at", "last_name", "first_name"]

    @property
    def full_name(self):
        return f"{self.first_name} {self.last_name}".strip()

    def __str__(self):
        return f"{self.full_name} ({self.email})"

    def save(self, *args, **kwargs):
        if not self.email:
            self.email = self.user.email
        super().save(*args, **kwargs)

    @transaction.atomic()
    def delete(self, *args, **kwargs):
        for child in self.children.all():
            if child.guardians.count() == 1:
                child.delete()
        return super().delete(*args, **kwargs)

    def user_can_view_contact_info(self, user: User) -> bool:
        return (
            self.user == user
            or user.has_global_permission(ProjectPermission.VIEW_FAMILIES)
            or any(
                user.can_view_families_in_project(child.project)
                for child in self.children.all()
            )
        )
