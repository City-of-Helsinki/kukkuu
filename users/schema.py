import graphene
from django.conf import settings
from django.contrib.auth import get_user_model
from django.db import transaction
from graphene import relay
from graphene_django import DjangoConnectionField
from graphene_django.types import DjangoObjectType
from graphql_jwt.decorators import login_required

from common.utils import update_object
from kukkuu.exceptions import KukkuuGraphQLError

from .models import Guardian

User = get_user_model()


LanguageEnum = graphene.Enum(
    "Language", [(l[0].upper(), l[0]) for l in settings.LANGUAGES]
)


class GuardianNode(DjangoObjectType):
    language = LanguageEnum()
    email = graphene.String()

    class Meta:
        model = Guardian
        interfaces = (relay.Node,)

    @classmethod
    def get_queryset(cls, queryset, info):
        return queryset.user_can_view(info.context.user).order_by("last_name")

    def resolve_email(self, info, **kwargs):
        return self.user.email


class UpdateMyProfileMutation(graphene.relay.ClientIDMutation):
    class Input:
        first_name = graphene.String()
        last_name = graphene.String()
        phone_number = graphene.String()
        language = LanguageEnum()

    my_profile = graphene.Field(GuardianNode)

    @classmethod
    @login_required
    @transaction.atomic
    def mutate_and_get_payload(cls, root, info, **kwargs):
        user = info.context.user

        try:
            guardian = user.guardian
        except Guardian.DoesNotExist as e:
            raise KukkuuGraphQLError(e)

        update_object(guardian, kwargs)

        return UpdateMyProfileMutation(my_profile=guardian)


class Query:
    guardians = DjangoConnectionField(GuardianNode)
    my_profile = graphene.Field(GuardianNode)

    @staticmethod
    @login_required
    def resolve_guardians(parent, info, **kwargs):
        return Guardian.objects.user_can_view(info.context.user)

    @staticmethod
    @login_required
    def resolve_my_profile(parent, info, **kwargs):
        return Guardian.objects.filter(user=info.context.user).first()


class Mutation:
    update_my_profile = UpdateMyProfileMutation.Field()
