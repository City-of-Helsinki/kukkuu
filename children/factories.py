import factory

from children.models import Child, Relationship
from projects.models import Project
from users.factories import GuardianFactory


class ChildFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    birthyear = factory.Faker(
        "random_int", min=2018, max=2023
    )  # NOTE: Many of the tests are freezed to 2020
    postal_code = factory.Faker("postcode", locale="fi_FI")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))
    notes = factory.Faker(
        "random_element",
        elements=[
            "",
            "Test notes",
            "Alternative test notes",
            "Longer test notes\nwith multiple lines.",
        ],
    )

    class Meta:
        model = Child


class RelationshipFactory(factory.django.DjangoModelFactory):
    child = factory.SubFactory(ChildFactory)
    guardian = factory.SubFactory(GuardianFactory)
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Relationship.TYPE_CHOICES]
    )

    class Meta:
        model = Relationship


class ChildWithGuardianFactory(ChildFactory):
    relationship = factory.RelatedFactory(RelationshipFactory, "child")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))


class ChildWithTwoGuardiansFactory(ChildWithGuardianFactory):
    relationship2 = factory.RelatedFactory(RelationshipFactory, "child")
