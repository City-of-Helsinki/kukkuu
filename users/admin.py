from django.contrib import admin

from children.models import Relationship
from users.models import Guardian


class RelationshipInline(admin.TabularInline):
    model = Relationship
    extra = 0
    fields = ("child", "type", "created_at")
    readonly_fields = ("created_at",)

    def has_change_permission(self, request, obj=None):
        return False


@admin.register(Guardian)
class GuardianAdmin(admin.ModelAdmin):
    inlines = (RelationshipInline,)
