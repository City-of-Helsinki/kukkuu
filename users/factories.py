import factory
from django.contrib.auth import get_user_model
from django.db.models import Q

from languages.models import Language
from users.models import Guardian


class UserFactory(factory.django.DjangoModelFactory):
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    username = factory.Faker("user_name")
    email = factory.Faker("email")

    class Meta:
        model = get_user_model()


class GuardianFactory(factory.django.DjangoModelFactory):
    user = factory.SubFactory(UserFactory)
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    phone_number = factory.Faker("phone_number")

    class Meta:
        model = Guardian

    @factory.post_generation
    def relationships(self, created, extracted, **kwargs):
        count = kwargs.pop("count", None)
        if count:
            from children.factories import RelationshipFactory

            RelationshipFactory.create_batch(count, guardian=self, **kwargs)

    @factory.post_generation
    def languages_spoken_at_home(self, created, extracted, **kwargs):
        if languages_spoken_at_home := extracted:
            filters = Q(alpha_3_code__in=languages_spoken_at_home)
            if None in languages_spoken_at_home:
                filters |= Q(alpha_3_code=None)
            self.languages_spoken_at_home.set(Language.objects.filter(filters))
