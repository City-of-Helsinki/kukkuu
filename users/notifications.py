from django.utils.translation import gettext_lazy as _
from django_ilmoitin.dummy_context import dummy_context
from django_ilmoitin.registry import notifications

from .factories import GuardianFactory


class NotificationType:
    GUARDIAN_EMAIL_CHANGED = "guardian_email_changed"


notifications.register(
    NotificationType.GUARDIAN_EMAIL_CHANGED, _("guardian email changed")
)

guardian = GuardianFactory.build()

dummy_context.update({NotificationType.GUARDIAN_EMAIL_CHANGED: {"guardian": guardian}})
