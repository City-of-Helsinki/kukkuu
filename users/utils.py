from django.contrib.auth import get_user_model
from django.core.exceptions import ValidationError
from django.core.validators import validate_email
from django_ilmoitin.utils import send_notification

from kukkuu.exceptions import InvalidEmailFormatError, VerificationTokenInvalidError
from users.notifications import NotificationType
from verification_tokens.models import VerificationToken

User = get_user_model()


def send_guardian_email_changed_notification(guardian):
    send_notification(
        guardian.email,
        NotificationType.GUARDIAN_EMAIL_CHANGED,
        context={"guardian": guardian},
        language=guardian.language,
    )


def validate_email_verification_token(user: User, verification_token_key: str):
    verification_token = VerificationToken.objects.filter(
        user=user,
        verification_type=VerificationToken.VERIFICATION_TYPE_EMAIL_VERIFICATION,
        key=verification_token_key,
    ).first()
    if not verification_token or not verification_token.is_valid():
        raise VerificationTokenInvalidError("The verification token is invalid")


def validate_guardian_email(email: str):
    try:
        validate_email(email)
    except ValidationError:
        raise InvalidEmailFormatError("Invalid email format")


def validate_guardian_data(guardian_data: dict):
    if "email" in guardian_data:
        validate_guardian_email(guardian_data["email"])
    return guardian_data
