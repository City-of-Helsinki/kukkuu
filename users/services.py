from typing import Optional, Union

from django.db.models import QuerySet
from django.template import Context, Template
from django_ilmoitin.utils import send_notification

from users.models import Guardian
from users.utils import get_communication_unsubscribe_ui_url


class GuardianEmailManagementNotificationService:
    @staticmethod
    def _send_notification(
        guardian: Guardian, notification_type: str, email: str, **kwargs
    ):
        context = kwargs.get("context", {})
        context["guardian"] = guardian
        context["unsubscribe_url"] = get_communication_unsubscribe_ui_url(
            guardian, guardian.language
        )
        send_notification(
            email, notification_type, context=context, language=guardian.language
        )

    @staticmethod
    def send_email_changed_notification(guardian: Guardian):
        """Send an email notification to guardian about a successful
        email change process. The email is now changed.

        Args:
            guardian (Guardian): the recipient guardian
        """
        from users.notifications import NotificationType

        GuardianEmailManagementNotificationService._send_notification(
            guardian, NotificationType.GUARDIAN_EMAIL_CHANGED, guardian.email
        )

    @staticmethod
    def send_email_update_token_notification(
        guardian: Guardian, email: str, verification_token_key: str
    ):
        """Send email update that contains the verification token key,
        that must be used to verify the new email during the change process.

        Args:
            guardian (Guardian): guardian whose email is being changed
            email (str): the new email address where the token is being sent
            verification_token_key (str): token to verify new email
        """
        from users.notifications import NotificationType

        GuardianEmailManagementNotificationService._send_notification(
            guardian,
            NotificationType.GUARDIAN_EMAIL_CHANGE_TOKEN,
            email,
            context={
                "verification_token": verification_token_key,
            },
        )


class AuthServiceNotificationService:
    @staticmethod
    def _send_auth_service_is_changing_notification(
        guardian: Guardian, date_of_change_str: Optional[str] = None
    ):
        """Send the auth service is changing notification to a guardian.

        Args:
            guardian (Guardian): the guardian instance
            date_of_change_str (Optional[str], optional): Date or datetime
                in string format. Defaults to None.
                The actual default should be given in the notification template.
        """
        from users.notifications import NotificationType

        send_notification(
            guardian.email,
            NotificationType.USER_AUTH_SERVICE_IS_CHANGING,
            context={
                "guardian": guardian,
                "date_of_change_str": date_of_change_str,
                "children_event_history_markdown": (
                    AuthServiceNotificationService.generate_children_event_history_markdown(  # noqa
                        guardian
                    )
                ),
            },
            language=guardian.language,
        )

    @staticmethod
    def send_user_auth_service_is_changing_notifications(
        guardians: Optional[Union[QuerySet, list[Guardian]]] = None,
        date_of_change_str: Optional[str] = None,
    ):
        """Send user authentication service is changing notifications
        to guariands as recipients.

        If the guardian list is not explicitly given
        the queryset result of
        `Guardian.objects.for_auth_service_is_changing_notification()`
        will be used as a default.

        Args:
            guardians (Optional[Union[QuerySet, list[Guardian]]], optional):
                explicit list of guardians as recipients. Defaults to None.
            date_of_change_str (Optional[str], optional): Date or datetime
                in string format. Defaults to None.
                The actual default should be given in the notification template.
        """
        if guardians is None:
            guardians = Guardian.objects.for_auth_service_is_changing_notification()

        for guardian in guardians:
            AuthServiceNotificationService._send_auth_service_is_changing_notification(
                guardian, date_of_change_str
            )

    @staticmethod
    def generate_children_event_history_markdown(guardian: Guardian):
        """Generates a Markdown string listing a guardian's children's enrolments."""

        # NOTE: This is markdown, so the line changes and white spaces are important!
        template_string = """
{% for child in guardian.children.all %}# {{ child.name }}
{% for enrolment in child.enrolments.all|dictsort:"occurrence.time" %}1. **{{ enrolment.occurrence.event.name }}:** {{ enrolment.occurrence.time|date:"j.n.Y H:i" }}{% if enrolment.occurrence.event.short_description %}
{{enrolment.occurrence.event.short_description}}{% endif %}
{% endfor %}
{% endfor %}
"""  # noqa

        template = Template(template_string)
        context = Context({"guardian": guardian})
        return template.render(context).strip()
