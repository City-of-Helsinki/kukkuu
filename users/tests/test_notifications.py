from unittest.mock import patch

import pytest
from django.core import mail

from children.factories import ChildWithGuardianFactory
from common.tests.conftest import create_api_client_with_user
from common.tests.utils import (
    assert_mails_match_snapshot,
    create_notification_template_in_language,
)
from events.factories import EnrolmentFactory
from users.factories import GuardianFactory
from users.notifications import NotificationType
from users.services import AuthServiceNotificationService
from users.tests.mutations import (
    REQUEST_EMAIL_CHANGE_TOKEN_MUTATION,
    UPDATE_MY_EMAIL_MUTATION,
)
from verification_tokens.factories import UserEmailVerificationTokenFactory


@pytest.fixture
def notification_template_guardian_email_changed_fi():
    return create_notification_template_in_language(
        NotificationType.GUARDIAN_EMAIL_CHANGED,
        "fi",
        subject="Guardian email changed FI",
        body_text="Guardian FI: {{ guardian }}. Unsubscribe: {{unsubscribe_url}}",
    )


@pytest.fixture
def notification_template_guardian_email_change_requested_fi():
    return create_notification_template_in_language(
        NotificationType.GUARDIAN_EMAIL_CHANGE_TOKEN,
        "fi",
        subject="Guardian email change verification token requested FI",
        body_text="""Guardian FI: {{ guardian }}.
        Token: {{ verification_token }}.
        Unsubscribe: {{unsubscribe_url}}
        """,
    )


USER_AUTH_SERVICE_IS_CHANGING_TEMPLATE = """
Guardian FI: {{ guardian }}.
The change is happening {{ date_of_change_str|default('17.6.2024', true) }}.
{% if guardian.children.count() %}
Childrens' event participation history:

{% for child in guardian.children.all() %}
Child name: {{child.name}}

{% for enrolment in child.enrolments.all() %}
Event: {{enrolment.occurrence.event.name}}
Occurrence: {{enrolment.occurrence.time}}

{% endfor %}
{% endfor %}
{% endif %}
{% if children_event_history_markdown %}
Markdown: 
{{children_event_history_markdown}}
{% endif %}"""  # noqa W291


@pytest.fixture
def notification_template_user_auth_service_is_changing_fi():
    return create_notification_template_in_language(
        NotificationType.USER_AUTH_SERVICE_IS_CHANGING,
        "fi",
        subject="User authorization service is changing FI",
        body_text=USER_AUTH_SERVICE_IS_CHANGING_TEMPLATE,
    )


@pytest.mark.parametrize(
    "new_email", ("new.email@example.com", "old.email@example.com", None)
)
@pytest.mark.django_db
def test_guardian_changed_email_notification(
    snapshot,
    new_email,
    notification_template_guardian_email_changed_fi,
    mock_user_create_subscriptions_management_auth_token,
):
    old_email = "old.email@example.com"
    guardian = GuardianFactory(email=old_email)
    verification_token = UserEmailVerificationTokenFactory(
        user=guardian.user, email=new_email
    )
    api_client = create_api_client_with_user(guardian.user)
    api_client.execute(
        UPDATE_MY_EMAIL_MUTATION,
        variables={
            "input": {"email": new_email, "verificationToken": verification_token.key}
        },
    )
    assert len(mail.outbox) == (1 if (new_email and new_email != old_email) else 0)
    assert_mails_match_snapshot(snapshot)


@patch("verification_tokens.models.VerificationToken.generate_key")
@pytest.mark.parametrize(
    "new_email", ("new.email@example.com", "old.email@example.com", None)
)
@pytest.mark.django_db
def test_guardian_change_email_token_requested_notification(
    mock_generate_key,
    new_email,
    snapshot,
    notification_template_guardian_email_change_requested_fi,
    mock_user_create_subscriptions_management_auth_token,
):
    # patch the return value for snapshot testing
    mock_generate_key.return_value = "abc123+-"
    old_email = "old.email@example.com"
    guardian = GuardianFactory(email=old_email)
    api_client = create_api_client_with_user(guardian.user)

    executed = api_client.execute(
        REQUEST_EMAIL_CHANGE_TOKEN_MUTATION,
        variables={"input": {"email": new_email}},
    )

    if new_email and new_email != old_email:
        assert (
            executed["data"]["requestEmailUpdateToken"]["emailUpdateTokenRequested"]
            is True
        )
        assert executed["data"]["requestEmailUpdateToken"]["email"] == new_email
    assert len(mail.outbox) == (1 if new_email else 0)
    assert_mails_match_snapshot(snapshot)


@pytest.mark.django_db
def test_send_user_auth_service_is_changing_notifications(
    notification_template_user_auth_service_is_changing_fi,
):
    GuardianFactory.create_batch(3, user__is_obsolete=True)
    AuthServiceNotificationService.send_user_auth_service_is_changing_notifications()
    assert len(mail.outbox) == 3


@pytest.mark.django_db
@pytest.mark.parametrize("date_of_change_str", [None, "", "24.12.2024"])
def test_send_user_auth_service_is_changing_with_date_of_change_str_param(
    date_of_change_str, snapshot, notification_template_user_auth_service_is_changing_fi
):
    """if no date_of_change_str is given, the default should be used.
    The default is set in `notification_template_user_auth_service_is_changing_fi`:
    "The change is happening 17.6.2024".
    """
    GuardianFactory(user__is_obsolete=True)
    AuthServiceNotificationService.send_user_auth_service_is_changing_notifications(
        date_of_change_str=date_of_change_str
    )
    assert len(mail.outbox) == 1
    assert_mails_match_snapshot(snapshot)


@pytest.mark.django_db
def test_send_user_auth_service_is_changing_with_children(
    snapshot, notification_template_user_auth_service_is_changing_fi
):
    guardian = GuardianFactory(user__is_obsolete=True)
    children = ChildWithGuardianFactory.create_batch(2, relationship__guardian=guardian)
    for child in children:
        EnrolmentFactory.create_batch(2, child=child)
    AuthServiceNotificationService.send_user_auth_service_is_changing_notifications()
    assert len(mail.outbox) == 1
    assert_mails_match_snapshot(snapshot)
