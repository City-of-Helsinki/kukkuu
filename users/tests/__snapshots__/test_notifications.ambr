# serializer version: 1
# name: test_guardian_change_email_token_requested_notification[None]
  list([
  ])
# ---
# name: test_guardian_change_email_token_requested_notification[new.email@example.com]
  list([
    '''
      kukkuu@example.com|['new.email@example.com']|Guardian email change verification token requested FI|Guardian FI: Michael Patton (old.email@example.com).
              Token: abc123+-.
              Unsubscribe: http://localhost:3000/fi/profile/subscriptions?authToken=a1b2c3d4e5f6g7h8
              
    ''',
  ])
# ---
# name: test_guardian_change_email_token_requested_notification[old.email@example.com]
  list([
    '''
      kukkuu@example.com|['old.email@example.com']|Guardian email change verification token requested FI|Guardian FI: Michael Patton (old.email@example.com).
              Token: abc123+-.
              Unsubscribe: http://localhost:3000/fi/profile/subscriptions?authToken=a1b2c3d4e5f6g7h8
              
    ''',
  ])
# ---
# name: test_guardian_changed_email_notification[None]
  list([
  ])
# ---
# name: test_guardian_changed_email_notification[new.email@example.com]
  list([
    "kukkuu@example.com|['new.email@example.com']|Guardian email changed FI|Guardian FI: Michael Patton (new.email@example.com). Unsubscribe: http://localhost:3000/fi/profile/subscriptions?authToken=a1b2c3d4e5f6g7h8",
  ])
# ---
# name: test_guardian_changed_email_notification[old.email@example.com]
  list([
  ])
# ---
